### ISSUE TITLE
Loom 1.1.2 on Windows does not release file locks at end of build

### ISSUE THREAD

**Issue by zml2008** [FabricMC/fabric-loom#813]
This appears to be a new regression in 1.1. Per the CI log from [adventure-platform-fabric](https://github.com/KyoriPowered/adventure-platform-fabric/actions/runs/3983153992/jobs/6828230484), it seems like a run of loom with an actively running daemon (aka not the first run, but subsequent runs) will always fail due to a held file lock.

I was able to reproduce this locally, though on my first try for some reason Loom found an existing cache lock. Honestly I'm not sure the best way to track down where the dangling file handle is being held -- my first guess would be some sort of JMX bean, but I didn't find anything there. I ended up taking a heap dump of the Gradle daemon JVM, and while I didn't see anything conclusive (and really, I'm not sure if a heap dump is the most effective way to do this investigation), I did see that some of the objects associated with the Mercury run were being retained -- something with shared services maybe?

**Comment by modmuss50** [comment]
Ill take a look, all my development and testing has only been done on MacOS so it does make it hard to spot issues like this.

In the past I have used https://github.com/jenkinsci/lib-file-leak-detector but last time I looked it did not support modern java, this may have changed as it seems to have active development. JProfiler is another good tool for detecting file handle leaks. 

> I did see that some of the objects associated with the Mercury run were being retained.

Maybe an issue with loom, however Mercury/JDT has had a long standing issue where it has a leak previously this was worked around by forcing a GC, maybe that workaround got lost in 1.1? As you mentioned in #779 using a worker for source remapping might help fully resolve this issue.

**Comment by modmuss50** [comment]
I was finally able to reproduce this and get something useful out of it using JProfiler. The project you provided is failing to compile for me, on the 2nd run I get the locked file issue.

After much effort in getting a profiler/debugger setup for this I was able to reproduce it, and got the following stracktrace for the File opened event:

```
org.gradle.api.internal.tasks.compile.AnnotationProcessingCompileTask.instantiateProcessor(java.lang.Class)
org.gradle.api.internal.tasks.compile.AnnotationProcessingCompileTask.setupProcessors()
org.gradle.api.internal.tasks.compile.AnnotationProcessingCompileTask.call()
org.gradle.api.internal.tasks.compile.ResourceCleaningCompilationTask.call()
org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(org.gradle.api.internal.tasks.compile.JavaCompileSpec)
org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(org.gradle.language.base.internal.compile.CompileSpec)
...
```
Full stacktrace: https://gist.github.com/modmuss50/2ea24d972f10793e9b2cc894856aa69a

I notice your build has a number of annotation processors, im not too sure how hard it would be for your to test without these, I wonder if any of them are keeping the file open after a failuire?I was unable to reproduce this with the example mod.

However that doesnt answer the question as to why this has only just started in loom 1.1 🤔 I will contiune to explore this, it seems it may be because `ProcessedNamedMinecraftProvider` is trying to regnerate the jar when it does not need to.

**Comment by modmuss50** [comment]
I have fixed and added many tests for the regression causing your build to fail @zml2008 in: https://github.com/FabricMC/fabric-loom/commit/a1e671b7190570e6ef9508d75f0aa9077ceb892c As I thought it was caused by the `ProcessedNamedMinecraftProvider` trying to recreate the jar file while being locked.

However this does not solve the root cause of the annotation processor keeping the file handle/lock open, I had a vuage memory of this being a issue in the past, and possibly the cause of most peoples frustration. I have perviously tried to always fork the java compiler in https://github.com/FabricMC/fabric-loom/commit/0cfa88f73b7b6da200dc5bac3dde3884a854e17f without success.

I imagine a compile and then a genSources is an easy way to reproduce this, its getting late now so I will come back to this another day.

**Comment by zml2008** [comment]
Thank you for looking into that -- apologies for the complexities, my personal vendetta against json is showing I suppose (the fabric.mod.json in that project is generated by a gradle task from a fabric.mod.yaml file, so anything like interface injection that reads the FMJ at configuration time doesn't work until the templates are generated, the daemon is stopped, and other build starts). At least that'll eventually go away once class tweaker is ready :)

I've been putting together a bit of a hunch while I was at work this afternoon after your comment about annotation processors -- I remembered that [Loom adds basically the entire compile classpath to the annotation processor path](https://github.com/FabricMC/fabric-loom/blob/e9362b75d9c9c5f43a38dcf4619acbfb92077722/src/main/java/net/fabricmc/loom/build/mixin/AnnotationProcessorInvoker.java#L143). This seems to happen since fabric-mixin-compile-extensions uses the runtime classpath (which is created from the annotation processor configuration in Gradle) to propogate mappings with its own logic.

Based on your profiler output, it seems like the annotation processor path (or at least its classloader) is being held on to here -- whether that's because of a misbehaving annotation processor not cleaning up its threads, or just the JVM earning its reputation of being pretty bad at releasing old class loaders (as we've learned via Bukkit's `/reload`). It appears that something in my project -- maybe it unconditionally setting a toolchain? is causing your fork option to not take effect (The handling of fork options is changing in Gradle 8 anyways, so imo it's something not worth looking at immediately). My thought here involved changing FMCE to walk the classpath using the AP's view of the compile classpath (via the Element (jx.processing)/TypeHandle (Mixin) model).

FMCE's behavor dates back to asie's original 2019 commit extracting the module from Loom. I'm not entirely sure their reasoning (since you could just as easily patch Mixin)? You may know better. It seems like Mixin has the logic to look up `TypeHandle`s and get their supertype information already in its remapping logic (`ObfuscationEnvironment#getObfMethod`), though Mixin's own handling does not process inherited field references -- I can't remember whether those are actually a thing in bytecode though. Mixin's handling also processes superclasses before implemented interfaces, the opposite of FMCE., and is subtly different in ways that may or may not impact functionality. Because it is subtly different, I didn't want to entirely nuke the FMCE interitence handling, so I've just migrated it to use the source model instead in FabricMC/fabric-mixin-compile-extensions#20.

With the FMCE pull request, plus updating the dep and commenting out the lines in AnnotationProcessorInvoker that [dump the whole compile classpath on the AP configuration](https://github.com/FabricMC/fabric-loom/blob/e9362b75d9c9c5f43a38dcf4619acbfb92077722/src/main/java/net/fabricmc/loom/build/mixin/AnnotationProcessorInvoker.java#L143), I am no longer able to consistently reproduce this issue.

I've confirmed that the hashes on the mixin mappings (in build/loom-cache) and the -dev jar are identical before and after these changes for my project -- though I'm sure there are loads of potential edge cases that would want further testing. (Unfortunately i can't compare the reobf jar since in whatever configuration Loom uses, TR does not produce bit-for-bit stable output).

That kinda gets me to the end of my evening today -- I'm working evenings tomorrow and the day after so the timezones may actually line up properly for us to be at a computer at the same time for once if you have a chance to look further.

**Comment by modmuss50** [comment]
> My thought here involved changing FMCE to walk the classpath using the AP's view of the compile classpath

This is great, I am not that familar with how the AP works, thanks for sharing your invaluable knowledge / making a PR.

> I'm not entirely sure their reasoning (since you could just as easily patch Mixin)? You may know better.

This is done to preseve support for upstream mixin, and also allows us to make fixes like this without a loader update. I think either way would be acceptable, this is just how we do it.

> TR does not produce bit-for-bit stable output

No it doesn't. this can be fixed by setting `preserveFileTimestamps` to false, and `reproducibleFileOrder` to true in the remap task. Maybe something we should enable by default, but thats for another issue.

I have gone ahead and merged your MCE PR after tesitng it with fabric API, I will get a loom PR up to allow the full suite of intergation tests to run. This appears to be a trivial fix but wouldnt be possible without your help/knowledge, thanks a lot for taking the time to help get this annoying issue finally fixed.
